kitsune/wiki/utils.py

```diff
@@ -133,28 +133,44 @@ def _active_contributors_id(from_date, to_date, locale, product):
     return set(list(editors) + list(reviewers))
 
 
-def get_featured_articles(product=None):
+def get_featured_articles(product=None, locale=settings.WIKI_DEFAULT_LANGUAGE):
     """Returns 4 random articles from the most visited.
 
     If a product is passed, it returns 4 random highly visited articles.
     """
-    featured = (
+    visits = (
         WikiDocumentVisits.objects.filter(period=LAST_7_DAYS)
         .exclude(
             document__products__slug__in=settings.EXCLUDE_PRODUCT_SLUGS_FEATURED_ARTICLES
         )
         .order_by("-visits")
-        .select_related("document")
+        .select_related('document')
     )
 
     if product:
-        featured = featured.filter(document__products__in=[product.id])
+        visits = visits.filter(document__products__in=[product.id])
+
+    documents = []
 
-    # Limit this to 10 entries
-    if featured.count() < 10:
-        return featured
+    if locale == settings.WIKI_DEFAULT_LANGUAGE:
+        for visit in visits:
+            documents.append(visit.document)
+            if len(documents) == 10:
+                break
+    else:
+        # prefretch localised documents to avoid n+1 problem
+        visits = visits.prefetch_related(
```

Comment:
What happens if there isn't a translation for the selected document? Shall we look for the next article in popularity to check for translations? Probably this should be continued until either we have a list or we exhaust our top ten. In that case we can either hide the articles for that locale or query for the next 10

Reply:
As written the `documents` array will be filled with the top 10 documents in any locale, ordered by their popularity in english.

Thinking about his over lunch, however, the prefetch query is going to be an absolute monster on the prod db, so we should limit the original query to 10 items. Testing it against the staging db I see a 100x speedup. I'll make this change and update the PR.

> In that case we can either hide the articles for that locale or query for the next 10

We only end up displaying 4 out of the 10, and I think it's highly unlikely that any locale won't have at least 4 of the 10 most popular articles in English localised. If they don't, then I'm not sure it's any more likely that the next 10 would... and then we're going to end up with a very expensive set of DB queries.